datalad_catalog/translate.py

```diff
@@ -0,0 +1,174 @@
+            & extractor_version_match
+            & schema_version_match
+        )
+
+    @abc.abstractmethod
+    def translate(self, metadata):
+        """
+        Translate incoming metadata to the current catalog schema
+
+        This method contains the logic for translating an incoming metadata
+        object to the catalog schema (version-specific).
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_supported_schema_version(self) -> str:
+        """
+        Report the version of the catalog schema supported by the translator
+
+        It will be matched to the `version` field in
+        `datalad_catalog/schema/jsonschema_catalog.json`.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_supported_extractor_name(self) -> str:
+        """
+        Report the name of the extractor supported by the translator
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_supported_extractor_version(self) -> str:
+        """
+        Report the version of the extractor supported by the translator
+
+        Use semantic versioning where possible
+        """
+        raise NotImplementedError
+
+    def get_current_schema_version(self) -> str:
+        """
+        Retrieves schema version of current package installation
+        """
+        # Setup schema parameters
+        package_path = Path(__file__).resolve().parent
+        schema_dir = package_path / "schema"
+        schema_path = schema_dir / str("jsonschema_catalog.json")
+        schema = read_json_file(schema_path)
+        return schema.get(cnst.VERSION)
+
+
+class TranslatorNotFoundError(InsufficientArgumentsError):
+    pass
+
+
+class Translate(object):
+    """
+    Class responsible for executing translation of an extracted metadata record,
+    including finding a matching class (base class TranslatorBase) and running
+    metadata translation.
+    """
+
+    def __init__(self, meta_record: dict = None) -> None:
+        """"""
+        self.meta_record = meta_record
+        if self.meta_record is None:
+            raise ValueError("No metadata record provided")
+        self.match_translator()
+
+    def match_translator(self):
+        """Match an extracted metadata record with an appropriate translator
+
+        Parameters
+        ----------
+        meta_record : dict
+            A dictionary containing metadata that was extracted from a file or dataset
+            using datalad-metalad
+
+        Returns
+        ------
+        TranslatorBase
+            The first matched translator instance of base class TranslatorBase
+        """
+        # First get source name and version from record
+        source_name = self.meta_record.get(cnst.EXTRACTOR_NAME)
+        source_version = self.meta_record.get(cnst.EXTRACTOR_VERSION)
+        matched_translators = []
+        for translator_name, translator_dict in get_translators().items():
```

Comment:
I'm not sure what the performance gain would be (or if there would be adverse effects), but considering that `match_translator()` (and `Translate()` itself) is meant to be called for a single line, I would consider calling `get_translators()` inside `datalad_catalog.catalog._translate_metadata()` and passing it to `Translate()` through an argument. This way we wouldn't need to iterate over entry points multiple times.

Reply:
good point, thanks!